routing/missioncontrol.go

```diff
@@ -219,9 +225,9 @@ func (m *missionControl) NewPaymentSession(routeHints [][]HopHint,
 					hopHint.FeeProportionalMillionths,
 				),
 				TimeLockDelta: hopHint.CLTVExpiryDelta,
+				Flags:         direction,
```

Comment:
Ideally this would be fixed in its own commit to make it clear what the consequence of this change is.

Reply:
The new test fails if the direction is not initialized here. I did try to add a unit test that directly checks this problem, but it didn't work as I expected; and I got into an impasse. 

I wanted the test to include two hints, one using a private channel in one direction and a second hint using the other direction of the same private channel. I wanted to have error on one of the directions, and check that the payment is routered through the other hint, through the private channel in a direction that didn't error. 

The setup was the following:  a graph including three nodes, a, b and c, connected by public edges in the following way.

               a  ------ b
               |
               |
               c

The payment should be routed to a node d. Node d is connected to b and c in private channels, and also c and b are connected in a private channel.

I included two hints   
h1: c -> b -> d
h2: b -> c -> d

I expected the route to be either 
r1 a -> c -> b -> d
r2 a -> b -> c -> d

But it seems that I'm getting either:
a1 a -> b -> d
a2 a -> c -> d

That is, I can't force routing to go through b -> c and c -> b, although that is what I explicitly indicate in the hints.
I was a bit surprised by that. Is routing expected to behave this way?

Initially I thought that I'd add negative fee on the b -- c channel to force routing to go through them, but unfortunately it doesn't seem that LND support negative fees at this point, as the fees are uint. 

I didn't have time to go deeper and make sure that I understand what is happening there correctly. I'll give it another try later.

Do you see a better test scenario?

Reply:
I documented my finding regarding the above attempted test here #2346. 

My conclusion is that up till now in all legit scenarios there was no need for adding direction to additionalEdges during initialization in NewPaymentSession. This was only really required when direction was considered in UpdateEdgePolicy (both in missioncontrol). 
I don't recommend removing the direction initialization and checking, it is the right thing to do, but it doesn't solve any real life scenario, i.e. all real life scenarios were ok if both were omitted. 

IMO this means that there was no 'subtle bug' found (or its so subtle its invisible) and
that initializing the direction in NewPaymentSession doesn't deserve its own commit and its own unit test. It is part of the other logic that needs to address the original bug this PR is supposed to fix. The unit test that was added test the directionality already (it fails without initialization of the direction in NewPaymentSession) covers it.

I can add a test that attempt to send a malformed onion error that reports the wrong direction of the reported edge and check that it doesn't update the additionalEdges, but it seems too contrived. 

Let me know what you think.
Thanks!

Reply:
Thanks for the detailed explanation! 

So if I understand it correctly, the reason this hasn't been a problem is because we don't this flag when doing path finding? We infer the direction of the edge by looking at the to/from nodes I think.

One way to possibly test this is to look at the `ignoredEdges` map. This one uses `edgeLocators` which also takes into account directionality. So if you could add the first direction to this map, then include a routing hint for the second direction, you might be able to make it fail finding a path.

Reply:
> So if I understand it correctly, the reason this hasn't been a problem is because we don't this flag when doing path finding? We infer the direction of the edge by looking at the to/from nodes I think.

Not exactly. What I'm saying is that if you didn't consider direction when pruning it'll be the same in real life scenarios.

Going back to the example detailed in the issue, suppose that we asked the router for 4 routes. 

We would get:
r1 A -> B -> D -> F
r2 A ->C -> E -> F
r3 A ->B -> D -> E -> F
r4 A ->C ->E -> D ->F

So it seems that we have both D->E and E->D on our list of candidates, and therefore we should make sure in code that if we receive an error on the D->E edge, we must prune only that direction D->E, but not E->D, hence the need for a direction check.

We should get an error on either directions of the E-D edge only once we attempt to route through it. We don't go through E-D in r1 and r2, so those need to fail first. 

So assume for our example that r1 failed due to D->F, so you prune it. and assume that r2 failed due to A->C, so you prune it too.

Now r3 is legit, so we try it. We pass through the D->E and we get an error.

With direction checking we prune D->E, but leave E->D. 

Without direction checking (as was before), we prune both E->D and D->E.

But r4 is useless anyhow (A->C already pruned), so it doesn't matter!

I'm quite confident that this can be formally proved for any graph whose cost at each hop is non-negative.

So bottom line, I don't see any real situation that I can simulate in a test that will require to check direction too, in addition to edge id, prior to pruning. I can create a contrived scenario, but as I said, the new test already test the direction checking anyhow.

Reply:
OK, I _think_ I see the problem you are facing testing this now. Here's a possible solution (using your original A-B-C-D graph):
* Include routing hints B-C-D and C-B-D
* Let the router try route A-B-C-D by making sure it is the cheapest path.
* Make the hop B-C fail, so it should be pruned.
* Now the router will pick either A-B-D or A-C-D, since it is impossible for A-C-B-D to be cheaper than both. Let's say it picks A-C-D. Make C-D fail.
* Now it picks A-B-D. Make A-B fail.
* The only possible route left is A-C-B-D. Before your bugfix this route wouldn't be found.

Reply:
> Let the router try route A-B-C-D by making sure it is the cheapest path.
That was my original plan.
But, I can't. The router will select A->C->D
That is why I opened the issue #2346

Reply:
Well, what if you make the fees on channels A-C and B-D super high? If it is still choosing the routes A-C-D or A-B-D then, you found a bug.

Reply:
Yes. I somehow convinced myself that its not possible. Anyhow, I separated this to a separate commit and added a dedicated unit test for that. Thanks!